File: handler.js
'use strict';

const fs = require('fs');
const twitter = require('twitter');
const crypto = require('crypto');

module.exports.main = async (event, context, callback) => {
  try {
    const header = event.headers ? event.headers['X-Hub-Signature'] || '' : '';
    const given = header.split('=')[1];
    const calc = crypto.createHmac('sha1', process.env.MIMIN_TOKEN).update(event.body).digest('hex');

    if (given !== calc) {
      console.log(`VALIDATION_ERROR: calc='${calc}', given='${given}'`);
      return callback(null, { statusCode: 400, body: "INVALID_TOKEN" });
    }

    const client  = new twitter({
      consumer_key:        process.env.MIMIN_TWITTER_COMSUMER_KEY,
      consumer_secret:     process.env.MIMIN_TWITTER_COMSUMER_SECRET,
      access_token_key:    process.env.MIMIN_TWITTER_ACCESS_TOKEN,
      access_token_secret: process.env.MIMIN_TWITTER_ACCESS_TOKEN_SECRET,
    });

    const param = JSON.parse(event.body);

    if (param.sender.login != "niltea") {
      throw Error("You are not mimin");
    }

    const hash  = param.after ? `(${param.after.substr(0,7)})` : "[unknown]";
    const image = fs.readFileSync('mimin_ga_mi.jpg');
    const media = await client.post('media/upload', { media: image });
    const ret   = await client.post('statuses/update', { status: `みみんがみ ${hash}`, media_ids: media.media_id_string });

    callback(null, { statusCode: 200, body: "OK" });
  } catch(err) {
    console.log("Error happen:", err.message);
    callback(null, { statusCode: 200, body: "ERROR" });
  }
};

File: handler.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/url"
	"os"

	"github.com/ChimeraCoder/anaconda"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
)

type GithubParam struct {
	Ref    string            `json:"ref"`
	Before string            `json:"before"`
	After  string            `json:"after"`
	Sender GithubParamSender `json:"sender"`
}

type GithubParamSender struct {
	Login string `json:"login"`
}

func CreateTwitterClient() *anaconda.TwitterApi {
	anaconda.SetConsumerKey(os.Getenv("MIMIN_TWITTER_COMSUMER_KEY"))
	anaconda.SetConsumerSecret(os.Getenv("MIMIN_TWITTER_COMSUMER_SECRET"))
	return anaconda.NewTwitterApi(os.Getenv("MIMIN_TWITTER_ACCESS_TOKEN"), os.Getenv("MIMIN_TWITTER_ACCESS_TOKEN_SECRET"))
}

func handler(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	var githubParam GithubParam
	json.Unmarshal([]byte(req.Body), &githubParam)

	if githubParam.Sender.Login != "niltea" {
		fmt.Println("You are not mimin")
		return events.APIGatewayProxyResponse{Body: "OK", StatusCode: 200}, nil
	}

	client := CreateTwitterClient()

	b, err := ioutil.ReadFile("mimin_ga_mi.jpg")

	if err != nil {
		return events.APIGatewayProxyResponse{Body: "OK", StatusCode: 200}, nil
	}

	media, err := client.UploadMedia(base64.StdEncoding.EncodeToString(b))

	if err != nil {
		return events.APIGatewayProxyResponse{Body: "OK", StatusCode: 200}, nil
	}

	p := url.Values{}
	p.Add("media_ids", media.MediaIDString)

	hash := githubParam.After[0:7]
	tweet := fmt.Sprintf("みみんがみ (%s)", hash)
	client.PostTweet(tweet, p)

	return events.APIGatewayProxyResponse{Body: "OK", StatusCode: 200}, nil
}

func main() {
	lambda.Start(handler)
}
